doc/source/train/user-guides/fault-tolerance.rst

```diff
@@ -5,128 +5,198 @@
-:ref:`saving <train-dl-saving-checkpoints>` *and* :ref:`loading checkpoints <train-dl-loading-checkpoints>`
-has been implemented.
+In these cases, the Ray Train driver (which calls ``trainer.fit()``) needs to be launched again.
+The relaunched Ray Train driver needs to find a minimal amount of run state in order to pick up where the previous run left off.
+This state includes the latest reported checkpoints, which are located at the :ref:`storage path <persistent-storage-guide>`.
+Ray Train fetches the latest checkpoint information from storage and passes it to the newly launched worker processes to resume training.
 
-.. literalinclude:: ../doc_code/dl_guide.py
-    :language: python
-    :start-after: __ft_initial_run_start__
-    :end-before: __ft_initial_run_end__
+To find this run state, Ray Train relies on passing in the **same** :class:`RunConfig(storage_path, name) <ray.train.RunConfig>` pair as the previous run.
+If the ``storage_path`` or ``name`` do not match, Ray Train will not be able to find the previous run state and will start a new run from scratch.
 
-The results and checkpoints of the experiment are saved to the path configured by :class:`~ray.train.RunConfig`.
-If the experiment has been interrupted due to one of the reasons listed above, use this path to resume:
+.. warning::
+    If ``name`` is reused unintentionally, Ray Train will fetch the previous run state, even if the user is trying to start a new run.
+    Therefore, always pass a unique run name when launching a new run. In other words, ``name`` should be a unique identifier for a training job.
 
-.. literalinclude:: ../doc_code/dl_guide.py
+.. note::
+    Job driver crashes and interrupts do not count toward the ``max_failures`` limit of :ref:`worker fault tolerance <train-worker-fault-tolerance>`.
+
+Here's an example training script that highlights best practices regarding job driver fault tolerance:
+
+.. literalinclude:: ../doc_code/fault_tolerance.py
     :language: python
-    :start-after: __ft_restored_run_start__
-    :end-before: __ft_restored_run_end__
+    :start-after: __job_driver_fault_tolerance_start__
+    :end-before: __job_driver_fault_tolerance_end__
 
-.. tip::
 
-    You can also restore from a remote path (e.g., from an experiment directory stored in a s3 bucket).
+Then, the entrypoint script can be launched with the following command:
 
-    .. literalinclude:: ../doc_code/dl_guide.py
-        :language: python
-        :dedent:
-        :start-after: __ft_restore_from_cloud_initial_start__
-        :end-before: __ft_restore_from_cloud_initial_end__
+.. code-block:: bash
 
-    .. literalinclude:: ../doc_code/dl_guide.py
-        :language: python
-        :dedent:
-        :start-after: __ft_restore_from_cloud_restored_start__
-        :end-before: __ft_restore_from_cloud_restored_end__
+    python entrypoint.py --storage_path s3://my_bucket/ --run_name unique_run_name-id=da823d5
```

Comment:
nit: `unique_run_name-id=da823d5` was kind of confusing to me, I thought this was setting a run ID at first. Maybe we can just call it `unique_run_name` here?

Also IDK if the user will have the wandb/mlflow run ID at this point, or if it would be generated inside the entrypoint script.

Reply:
If you want to connect to the same wandb run upon the job retry, you should set the run id on the outside and pass it in, `wandb.init(..., run_id=run_id)`. Otherwise it's randomly generated.

I am setting a run id here as a best practice for the user submitting a job.